[OMPIRBuilder] Avoid crash in BasicBlock::splice.

abidh opened the pull request:
Calling `BasicBlock::splice` in `spliceBB` when both `Old` and `New` are empty is a `nop` currently but it can cause a crash once debug records are used instead of debug intrinsics. This PR makes the call conditional on at least one of `Old` or `New` being non-empty.

Consider the following mlir:
```
omp.target map_entries() {
  llvm.intr.dbg.declare ...
  llvm.intr.dbg.declare ...
  omp.teams ...
  ...
}
```

Current code would translate llvm.intr Ops to llvm intrinsics. Old is the BasicBlock where they were get inserted and it will have 2 llvm debug intrinsics by the time the implementation of `omp.teams`  starts. This implementation creates many BasicBlocks by calling `splitBB`. The `New` is the just created BasicBlock which is empty.

In the new scheme (using debug records), there will be no instruction in the `Old` BB after llvm.intr Ops get translated but just 2 trailing debug records. So both `Old` and `New` are empty. When control reaches `BasicBlock::splice`, it calls `spliceDebugInfoEmptyBlock`. This function expects that in this case (`Src` is empty but has trailing debug records), the `ToIt` is valid and it can call `adoptDbgRecords` on it. This assumption is not true in this case as `New` is empty and `ToIt` is pointing to end(). The fix is to only call `BasicBlock::splice` when at least of `Old` or `New` is not empty.

abidh committed:
[OMPIRBuilder] Avoid crash in BasicBlock::splice.

Calling splice when both Old and New are empty is a nop currently but
it can cause a crash once debug records are used instead of debug
intrinsics. This PR makes the call conditional on at least one of
Old or New being non-empty.

Consider the following mlir:
omp.target map_entries() {
  llvm.intr.dbg.declare ...
  llvm.intr.dbg.declare ...
  omp.teams ...
  ...
}

Current code would translate llvm.intr to llvm intrinsics and we will
have 2 of them in the Old BB by the time omp.teams implementation starts.
This implementation creates many BasicBlocks by calling splitBB.

In the new scheme (using debug records), there will be no instruction
in the Old BB after llvm.intr get translated but just 2 trailing debug
records. So effectively both Old and New are empty. When control reaches
BasicBlock::splice, it calls spliceDebugInfoEmptyBlock. This funtion
expects that in this case (Src is empty but has trailing debug records),
the ToIt is valid and it can call adoptDbgRecords on it. This
assumption is not true in this case as New is empty and ToIt is pointing
to end().

Meinersbur commented:
I don't follow what is happening. Why doesn't `BasicBlock::splitBasicBlock` (where `llvm::spliceBB` was inspired by) need this check?

Can you add a unittest for illustration? I don't mean a translation from MLIR, but some IRBuilder-generated LLVM-IR on which `llvm::spliceBB` is called in `OpenMPIRBuilderTest.cpp`

abidh committed:
Add a unit test.

Meinersbur commented:
Thank you for the unittest, it helps me understand the issue.

It seems that `spliceDebugInfoEmptyBlock` wants to move "trailing DbgRecords" to the last instruction of the destination block. Since it is degenerate, such last instruction does not exist. 

I think that `spliceDebugInfoEmptyBlock` is at fault here. If it wants to move "trailing DbgRecords" to an empty Dest, then it should append those DebugRecords also to the trailing DbgRecords of Dest (setTrailingDbgRecords).

abidh commented:
I agree with your point that moving dbg records from `Old` to `New` only when New is not empty is inconsistent. Would it work if we only checked that `Old` is empty. That will remove the inconsistency.

I am a bit hesitant about adding `UnreachableInst` as change may be required in multiple places and lifetime may be long. So I was a bit concerned with any side effects.

abidh committed:
Avoid splice call only when Old is empty.

This removes the inconsistent behavior in the PR. Also added comments to describe the rationale.

Meinersbur reviewed (approved):
LGTM

abidh committed:
Update comments.
